### ISSUE TITLE
Cancel button moves in downloading forms dialog

### ISSUE THREAD

**Issue by dbemke** [getodk/collect#5711]
#### ODK Collect version
the store version 2023.2.3, the master version ce19503cb216b93eb558bc9187211551a64cb447
#### Android version
10, 13
#### Device used
Redmi 9T, Galaxy A23
#### Problem description
The dialog which appears during downloading forms changes its size so the Cancel button moves and makes it more difficult to tap the button.
Video - cancelbutton file on https://drive.google.com/drive/u/0/folders/1IeuxQZU8Sw90PAOnTRlMToB3Wp_LuCoN
#### Steps to reproduce the problem
1. Go to Download form.
2. Select a few form to download.
3. Tap Get Selected.
4. Try to tap Cancel in the dialog.
#### Expected behavior
Cancel button shouldn’t move/ dialog shouldn't change its size.

**Comment by manojkumar20k3** [comment]
Hello, I am a beginner in open source contribution. Is this issue open ? If so please assign it to me ,So that i can make my first contribution .

**Comment by seadowg** [comment]
Hi @manojkumar20k3! This issue is not ready for work yet. If you'd like to make a contribution, reading our [contribution guide](https://github.com/getodk/collect/blob/master/docs/CONTRIBUTING.md) would be a good starting point.

**Comment by seadowg** [comment]
@dbemke the video file isn't accessible

**Comment by seadowg** [comment]
Managed to get to the video. Thanks @dbemke!

This is definitely not ideal. There are a few ways to fix this that I can see:

- Limit the max lines for the statuses in the progress dialog - this would truncate longer statuses on thinner screens
- Move the dynamic statuses below the spinner indicator (show fixed text next to the indicator like "Downloading forms...") and limit the max lines - this would still risk truncation, but would give us more room before it happens


@alyblenkin do you have any ideas for how to improve this interaction?

**Comment by alyblenkin** [comment]
> @alyblenkin do you have any ideas for how to improve this interaction?

I'm curious to know what we think is the most helpful information to display because there is a lot to read and it happens so fast. I'm assuming the user goal is to know that the forms they've selected are actively downloading.

If that's the case, I'm wondering if we can simplify the progress dialog? Do we need to say "Connecting to the server" and each individual item that is downloading? Could we just have the spinner and "Downloading 1/7 forms"?

**Comment by seadowg** [comment]
> If that's the case, I'm wondering if we can simplify the progress dialog? Do we need to say "Connecting to the server" and each individual item that is downloading? Could we just have the spinner and "Downloading 1/7 forms"?

I think for forms without media files this makes sense. When we do have media files however, each form could take a while on a slow connection (imagine a form with a lot of images/audio being downloaded on a 2G connection). I'd lean towards still including the progress for media downloads as well.

**Comment by alyblenkin** [comment]
> I think for forms without media files, this makes sense. When we do have media files, however, each form could take a while on a slow connection (imagine a form with a lot of images/audio being downloaded on a 2G connection). I'd lean towards still including the progress for media downloads as well.

Yeah that's a good point; with a slower connection you would want more feedback. Another option is showing a determinant indicator that [tracks the progress](https://m3.material.io/components/progress-indicators/guidelines#9e82c46f-ed52-4642-b429-5bf505311462) if we know how long it will take.

We could also display the progress indicator at the top of the dialog, above the header, so we are not creating gaps and avoid truncating information. [Quick mockup](https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3346%3A90744&mode=design&t=e8qtyX5T8rTo9akw-1) 

Another thing for us to consider down the road is allowing users to continue doing another task in parallel if we estimate it to take longer to download. Or at least show the percentage or estimated time it will take to download. I'll note this down for later :)

**Comment by seadowg** [comment]
> We could also display the progress indicator at the top of the dialog, above the header, so we are not creating gaps and avoid truncating information. [Quick mockup](https://www.figma.com/file/KFi9hIQdrRgo0rLqIC7Sg8/ODK-user-flows?type=design&node-id=3346%3A90744&mode=design&t=e8qtyX5T8rTo9akw-1)

That's a really nice tweak. We'd still need line limits, but they'd be far less problematic. We be happy with all progress dialogs changing to look like that?

**Comment by alyblenkin** [comment]
@seadowg - agreed, we will still need line limits, but hopefully it makes it easier to read! Changing the dialog title to "Downloading forms" if there are multiple or "Downloading form" if there is only one would also be more user-friendly. This might be a different task. 

I also thought about making this change to all the dialogs last night! I don't see any downsides to making that change. Moving the progress indicator makes it easier to scan and understand what's happening.

**Comment by lognaturel** [comment]
> it happens so fast

Really important to remember that even very small forms without media files can take a long time to download! The only reason they're so fast for us is because we are very close to where they're served from and have ridiculously fast connections by global standards.

Should we perhaps consider looking into ways to emulate a slow connection so we get a more realistic experience? A quick search shows me https://play.google.com/store/apps/details?id=com.bwr.free&hl=en_US&gl=US, maybe there are other good options you all know about?

**Comment by seadowg** [comment]
> Should we perhaps consider looking into ways to emulate a slow connection so we get a more realistic experience? A quick search shows me https://play.google.com/store/apps/details?id=com.bwr.free&hl=en_US&gl=US, maybe there are other good options you all know about?

It's something I often play around with in an emulator when I'm adding progress indicators. There's "Network Type" and "Signal Strength" settings under "Cellular" in the "Extended Controls" (the options menu).

**Comment by alyblenkin** [comment]
Yes, it's a great point. I haven't played around with the bandwidth ruler, but I'll check it out. I've been thinking about different types of loading and progress indicators we may want to use to give the user more appropriate feedback. 

I mentioned above that we might want to consider a different type of indicator (e.g. show the percentage or estimated time it will take to download). I don't know if we can determine the wait time, but it could be helpful feedback. Is it worth considering it sooner? It aligns with our transparency principle and usability heuristic around giving [visibility of system status](https://www.nngroup.com/articles/ten-usability-heuristics/) to create more trust. 

I came across this [great article](https://uxdesign.cc/loading-progress-indicators-ui-components-series-f4b1fc35339a) that outlines a few options when selecting the best way to give user feedback so they stay informed during wait times. 

![image](https://github.com/getodk/collect/assets/7871340/a503be69-1367-489f-a3c7-bc3485cd4e54)

**Comment by seadowg** [comment]
We're going to focus on work to the "Exactly Match Server" flows, so don't plan on improving this any time soon.
